File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char,char> mp = {{'(',')'},{'[',']'},{'{','}'}};

        for (char c : s) {
            //if it's an opening bracket, push
            if (mp.count(c)) {
                st.push(c);
            } 
            //if it's a closing bracket, check match
            else {
                if (st.empty() || mp[st.top()] != c) return false;
                st.pop();
            }
        }

        return st.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mp = {'(': ')', '[': ']', '{': '}'}

        for c in s:
            # if it's an opening bracket, push
            if c in mp:
                stack.append(c)
            # if it's a closing bracket, check match
            else:
                if not stack or mp[stack[-1]] != c:
                    return False
                stack.pop()
        
        return not stack
